### PR TITLE
expression, executor: rewrite builtin func Cos

### PR DESCRIPTION
**Pull request by plutolove**

### CODE REVIEW

File: expression/builtin_math_test.go [hunk 1]
```diff
@@ -561,29 +562,43 @@ func (s *testEvaluatorSuite) TestSin(c *C) {
...
+		isNil    bool
+		getErr   bool
 	}{
-		{nil, nil},
-		{int64(0), float64(1)},
-		{math.Pi, float64(-1)}, // cos pi equals -1
-		{-math.Pi, float64(-1)},
-		{math.Pi / 2, float64(math.Cos(math.Pi / 2))}, // Pi/2 is some near 0 (6.123233995736766e-17) but not 0. Even in math it is 0.
-		{-math.Pi / 2, float64(math.Cos(-math.Pi / 2))},
-		{"0.000", float64(1)}, // string value case
```
Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: why remove this case?

File: expression/builtin_math.go [hunk 2]
```diff
@@ -843,38 +844,36 @@ func (b *builtinAtanSig) eval(row []types.Datum) (d types.Datum, err error) {
...
 
+//cos start
 type cosFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *cosFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinCosSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeDouble)
+	tp.Flen = 8
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: According to [MySQL source code](https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.cc#L2813), we may set it to 23 (DBL_DIG + 8)

File: expression/builtin_math_test.go (hunk 1 above)
Comment (plutolove) [inline expression/builtin_math_test.go]: just rewrite

Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: we may keep this as valid string input argument. @plutolove

Comment (plutolove) [inline expression/builtin_math_test.go]: ok , I fix it right now

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by winoros** [comment]
LGTM @plutolove please resolve conficts.

File: expression/builtin_math.go [hunk 3]
```diff
@@ -843,38 +844,36 @@ func (b *builtinAtanSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return d, nil
 }
 
+//cos start
```
Comment (lamxTyler) [inline expression/builtin_math.go]: Why add this comment?

Comment (plutolove) [inline expression/builtin_math.go]: Flag code range of builtin-cos

Comment (lamxTyler) [inline expression/builtin_math.go]: There is no other place in this file has this kind of comment, so I suggest that this can be removed.

Comment (plutolove) [inline expression/builtin_math.go]: ok, I got it

**Comment by stdupp** [comment]
If the arg is a string with a valid float as prefix like "1.1xxx", it seems doesn't work. In MySQL sin("1.1xxx") will return 0.8912073600614354

**Comment by winoros** [comment]
@stdupp evalReal will do the type cast.

**Comment by stdupp** [comment]
@winoros But in builtinCastStringAsRealSig.evalReal we use 'strconv.ParseFloat' not 'types.StrToFloat'

**Comment by XuHuaiyu** [comment]
@stdupp Yeah, it's a mistake. I'll fix it.

**Pull request merged**